lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_firmware.py

```diff
@@ -0,0 +1,228 @@
+
+
+def update_firmware(idrac, module):
+    """Update firmware from a network share and return the job details."""
+    msg = {}
+    msg['changed'] = False
+    msg['msg'] = {}
+    err = False
+
+    try:
+        invalid, message = _validate_catalog_file(module.params)
+        if invalid:
+            err = True
+            msg['msg'] = message
+            return msg, err
+
+        upd_share = FileOnShare(remote=module.params['share_name'] + "/" + module.params['catalog_file_name'],
+                                mount_point=module.params['share_mnt'],
+                                isFolder=False,
+                                creds=UserCredentials(
+                                    module.params['share_user'],
+                                    module.params['share_pwd'])
+                                )
+
+        idrac.use_redfish = True
+        if '12' in idrac.ServerGeneration or '13' in idrac.ServerGeneration:
+            idrac.use_redfish = False
+
+        apply_update = True
+        msg['msg'] = idrac.update_mgr.update_from_repo(upd_share,
+                                                       apply_update,
+                                                       module.params['reboot'],
+                                                       module.params['job_wait'])
+
+        if "Status" in msg['msg']:
+            if msg['msg']['Status'] == "Success":
+                if module.params['job_wait'] is True:
+                    msg['changed'] = True
+            else:
+                err = True
+
+    except Exception as e:
+        err = True
+        msg['msg'] = "Error: %s" % str(e)
+
+    return msg, err
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+
+            # iDRAC Credentials
+            idrac_ip=dict(required=True, type='str'),
+            idrac_user=dict(required=True, type='str'),
+            idrac_pwd=dict(required=True, type='str', no_log=True),
+            idrac_port=dict(required=False, default=443, type='int'),
+
+            # Network File Share
+            share_name=dict(required=True, type='str'),
+            share_user=dict(required=False, type='str'),
+            share_pwd=dict(required=False, type='str', no_log=True),
+            share_mnt=dict(required=True, type='str'),
+
+            # Firmware update parameters
+            catalog_file_name=dict(required=False, type='str', default='Catalog.xml'),
+            reboot=dict(required=False, default=False, type='bool'),
+            job_wait=dict(required=False, default=True, type='bool')
+        ),
+
+        supports_check_mode=False)
+
+    if not HAS_OMSDK:
+        module.fail_json(msg="Dell EMC OpenManage Python SDK required for this module")
+
+    # Connect to iDRAC
+    idrac_conn = iDRACConnection(module)
+    idrac = idrac_conn.connect()
+
+    msg, err = update_firmware(idrac, module)
+
+    # Disconnect from iDRAC
+    idrac_conn.disconnect()
+
+    if err:
```

Comment:
One of the things you could do is to have a custom exception hierarchy with some failure details attached to it and then try/except for that in just one place. Then if your main processing/action function returns successfully use that data to `exit_json`, otherwise use data attached to exception in order to `fail_json`. It'd be one place for returning such data.

Here's an example of how I used such approach myself:
https://github.com/webknjaz/ansible-gentoo-laptop/blob/master/library/eix.py#L121-L128

Reply:
updated with `fail_json`